rust/arrow/src/compute/kernels/take.rs

```diff
@@ -0,0 +1,584 @@
+        DataType::List(_) => take_list(values, indices),
+        DataType::Struct(fields) => {
+            let struct_: &StructArray =
+                values.as_any().downcast_ref::<StructArray>().unwrap();
+            let arrays: Result<Vec<ArrayRef>> = struct_
+                .columns()
+                .iter()
+                .map(|a| take(a, indices, Some(options.clone())))
+                .collect();
+            let arrays = arrays?;
+            let pairs: Vec<(Field, ArrayRef)> =
+                fields.clone().into_iter().zip(arrays).collect();
+            Ok(Arc::new(StructArray::from(pairs)) as ArrayRef)
+        }
+        t @ _ => unimplemented!("Take not supported for data type {:?}", t),
+    }
+}
+
+/// Options that define how `take` should behave
+#[derive(Clone)]
+pub struct TakeOptions {
+    /// Perform bounds check before taking indices from values.
+    /// If enabled, an `ArrowError` is returned if the indices are out of bounds.
+    /// If not enabled, and indices exceed bounds, the kernel will panic.
+    pub check_bounds: bool,
+}
+
+impl Default for TakeOptions {
+    fn default() -> Self {
+        Self {
+            check_bounds: false,
+        }
+    }
+}
+
+/// `take` implementation for primitive arrays
+fn take_primitive<T>(values: &ArrayRef, indices: &UInt32Array) -> Result<ArrayRef>
+where
+    T: ArrowPrimitiveType,
+{
+    let mut builder = PrimitiveBuilder::<T>::new(indices.len());
+    let a = values.as_any().downcast_ref::<PrimitiveArray<T>>().unwrap();
+    for i in 0..indices.len() {
+        if indices.is_null(i) {
+            builder.append_null()?;
+        } else {
+            let ix = indices.value(i) as usize;
+            if a.is_valid(ix) {
+                builder.append_value(a.value(ix))?;
+            } else {
+                builder.append_null()?;
+            }
+        }
+    }
+    Ok(Arc::new(builder.finish()) as ArrayRef)
+}
+
+/// `take` implementation for binary arrays
+fn take_binary(values: &ArrayRef, indices: &UInt32Array) -> Result<ArrayRef> {
+    let mut builder = BinaryBuilder::new(indices.len());
+    let a = values.as_any().downcast_ref::<BinaryArray>().unwrap();
+    for i in 0..indices.len() {
+        if indices.is_null(i) {
+            builder.append(false)?;
+        } else {
+            let ix = indices.value(i) as usize;
+            if a.is_null(ix) {
+                builder.append(false)?;
+            } else {
+                builder.append_value(a.value(ix))?;
+            }
+        }
+    }
+    Ok(Arc::new(builder.finish()) as ArrayRef)
+}
+
+/// `take` implementation for list arrays
+///
+/// Calculates the index and indexed offset for the inner array,
+/// applying `take` on the inner array, then reconstructing a list array
+/// with the indexed offsets
+fn take_list(values: &ArrayRef, indices: &UInt32Array) -> Result<ArrayRef> {
+    // TODO Some optimizations can be done here such as if it is taking the whole list or a contiguous sublist
```

Comment:
nit: let's fix the format: `TODO Some` -> `TODO: some`. Also break this long line into two.